src/flash.rs

```diff
@@ -2,6 +2,242 @@
 
 use crate::pac::{flash, FLASH};
 
+pub const FLASH_START: u32 = 0x0800_0000;
+pub const FLASH_END:   u32 = 0x080F_FFFF;
+
+const RDPRT_KEY: u16 = 0x00A5;
+const KEY1: u32 = 0x45670123;
+const KEY2: u32 = 0xCDEF89AB;
+
+pub const SZ_1K: u16 = 1024;
+
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Error {
+    Success,
+    AddressLargedThanFlash,
+    AddressMisAligned,
+    LengthNotMultiple2,
+    LengthTooLong,
+    EraseError,
+    ProgrammingError,
+    WriteError,
+    VerifyError,
+    UnLockError,
+    LockError,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum SectorSize {
+    Sz1K = 1,
+    Sz2K = 2,
+    Sz4K = 4,
+}
+impl SectorSize {
+    fn size(self) -> u16 {
+        SZ_1K * self as u16
+    }
+}
+
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum FlashSize {
+    Sz16K  = 16,
+    Sz32K  = 32,
+    Sz64K  = 64,
+    Sz128K = 128,
+    Sz256K = 256,
+    Sz384K = 384,
+    Sz512K = 512,
+    Sz768K = 768,
+    Sz1M   = 1024,
+}
+impl FlashSize {
+    fn size(self) -> u32 {
+        SZ_1K as u32 * self as u32
+    }
+}
+
+pub struct FlashWriter<'a> {
+    flash: &'a mut Parts,
+    sector_sz: SectorSize,
+    flash_sz: FlashSize,
+}
+impl<'a> FlashWriter<'a> {
+    fn unlock(&mut self) -> Result<()> {
+        // Wait for any ongoing operations
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // NOTE(unsafe) write Keys to the key register
+        unsafe { self.flash.keyr.keyr().write(|w| w.key().bits(KEY1)); }
+        unsafe { self.flash.keyr.keyr().write(|w| w.key().bits(KEY2)); }
+
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Verify success
+        match self.flash.cr.cr().read().lock().bit_is_clear() {
+            true => Ok(()),
+            false => Err(Error::UnLockError),
+        }
+    }
+
+    fn lock(&mut self) -> Result<()> {
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        self.flash.cr.cr().modify(|_, w| w.lock().set_bit());
+
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        match self.flash.cr.cr().read().lock().bit_is_set() {
+            true => Ok(()),
+            false => Err(Error::LockError),
+        }
+    }
+
+    fn valid_address(&self, offset: u32) -> Result<()> {
+        if FLASH_START + offset > FLASH_END {
+            Err(Error::AddressLargedThanFlash)
+        }
+        else if offset % 2 != 0 {
```

Comment:
I wonder if the compiler will always do the right thing and change this to `offset & 1`, otherwise it might be better if we use bitwise AND instead of mod.

Reply:
I had a look at the LLVM-IR produced by rustc, and it looks like it doesn't change it to an `and` in debug mode, but does in release mode. That is without LLVM optimizations though